### PR TITLE
Add CarthageDEX 

### PR DESCRIPTION
**Pull request by samisbakedham**
**NOTE**

1. The protocol is usually listed within 24 hours of merging the PR
2. Please fill the form below  **only if the PR is for listing a new protocol** else it can be ignored/replaced with reason/details about the PR
3. **For updating listing info** It is a different repo, you can find your listing in this file: https://github.com/DefiLlama/defillama-server/blob/master/defi/src/protocols/data2.ts, you can  edit it there and put up a PR
4. Do not edit/push `package-lock.json` file as part of your changes, we use lockfileVersion 2, and most use v1 and using that messes up our CI

---

##### Short Description (to be shown on DefiLlama):
CarthageDEX is a fork of uniswap deployed on Candle.

##### Category (full list at https://defillama.com/categories) *Please choose only one:
DEX

##### Oracle used (Chainlink/Band/API3/TWAP or any other that you are using):
Chainlink

##### forkedFrom (Does your project originate from another project):
Uniswap

##### methodology (what is being counted as tvl, how is tvl being calculated):
All liquidity provided to pools

### CODE REVIEW

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    Type `Query` has no field `carthageFactory`
    Truncated error stack:
        at /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/graphql-request/dist/index.js:340:31

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    Type `Query` has no field `carthageFactory`
    Truncated error stack:
        at /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/graphql-request/dist/index.js:340:31

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    Type `Query` has no field `carthageFactory`
    Truncated error stack:
        at /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/graphql-request/dist/index.js:340:31

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    Type `Query` has no field `uniswapFactory`
    Truncated error stack:
        at /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/graphql-request/dist/index.js:340:31

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    Invalid value provided for argument `block`: Object({"number": Null})
    Truncated error stack:
        at /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/graphql-request/dist/index.js:340:31

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    Invalid value provided for argument `block`: Object({"number": Null})
    Truncated error stack:
        at /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/graphql-request/dist/index.js:340:31

**Comment by llamatester** [comment]
The adapter at projects/carthage exports TVL: 
        
 
 
    candle                    0
    
    total                    0

**Comment by g1nt0ki** [comment]
@samisbakedham hi, can you fill in the form with details about your project like what the project site/twiiter handle/etc?

**Comment by g1nt0ki** [comment]
Also, is this an uniswap fork? is this the correct factory address? 0x5Bb7BAE25728e9e51c25466D2A15FaE97834FD95

**Comment by samisbakedham** [comment]
> Also, is this an uniswap fork? is this the correct factory address? 0x5Bb7BAE25728e9e51c25466D2A15FaE97834FD95

Hi,
Yes this is a uniswap fork. That is the correct factory address as well!

**Comment by samisbakedham** [comment]
> @samisbakedham hi, can you fill in the form with details about your project like what the project site/twiiter handle/etc?

Will do

**Comment by g1nt0ki** [comment]
> > Also, is this an uniswap fork? is this the correct factory address? 0x5Bb7BAE25728e9e51c25466D2A15FaE97834FD95
> 
> Hi, Yes this is a uniswap fork. That is the correct factory address as well!

hmm, when I tried making 'allPairsLength' api call to that contract it fails, do you use a public rpc, do you mind sharing? wondering if it is an rpc issue but I doubt it.

**Comment by samisbakedham** [comment]
> > > Also, is this an uniswap fork? is this the correct factory address? 0x5Bb7BAE25728e9e51c25466D2A15FaE97834FD95
> > 
> > 
> > Hi, Yes this is a uniswap fork. That is the correct factory address as well!
> 
> hmm, when I tried making 'allPairsLength' api call to that contract it fails, do you use a public rpc, do you mind sharing? wondering if it is an rpc issue but I doubt it.

The RPC for candle is https://rpc.cndlchain.com

**Comment by g1nt0ki** [comment]
@samisbakedham the site is down?

**Comment by samisbakedham** [comment]
> @samisbakedham the site is down?

Typo, just fixed it

**Comment by g1nt0ki** [comment]
@samisbakedham I was trying your graph endpoint, the token count in pools are correct, but prices are like really off, like $122 for cndl token, are you looking into it?


    query get_tvl {
      
      pools {
        id
        token0 {
          id
        }
        token1 {
          id
        }
        liquidity
        sqrtPrice
        token0Price
        token1Price
        totalValueLockedUSD
        totalValueLockedETH
        totalValueLockedToken0
        totalValueLockedToken1
      }
    }

File: speedTest.js [hunk 1]
```diff
@@ -21,6 +21,7 @@ function createProvider(name, url, chainId) {
   const providers = {
     ethereum: createProvider("ethereum", "https://mainnet.infura.io/v3/9aa3d95b3bc440fa88ea12eaa4456161", 1),
     bsc: createProvider("bsc", "https://bsc-dataseed4.binance.org", 56),
+    candle: createProvider("candle", "https://candlelabs.org", 534),
```
Comment (g1nt0ki) [inline speedTest.js:24]: please revert changes to this file

Comment (samisbakedham) [reply speedTest.js:24]: Is there a reason?

**Comment by samisbakedham** [comment]
> @samisbakedham I was trying your graph endpoint, the token count in pools are correct, but prices are like really off, like $122 for cndl token, are you looking into it?
> 
> ```
> query get_tvl {
>   
>   pools {
>     id
>     token0 {
>       id
>     }
>     token1 {
>       id
>     }
>     liquidity
>     sqrtPrice
>     token0Price
>     token1Price
>     totalValueLockedUSD
>     totalValueLockedETH
>     totalValueLockedToken0
>     totalValueLockedToken1
>   }
> }
> ```

image
122 WCNDL to one USDC

File: speedTest.js (hunk 1 above)
Comment (samisbakedham) [reply speedTest.js:24]: My bad! Thank you

**Comment by g1nt0ki** [comment]
> > @samisbakedham I was trying your graph endpoint, the token count in pools are correct, but prices are like really off, like $122 for cndl token, are you looking into it?
> > ```
> > query get_tvl {
> >   
> >   pools {
> >     id
> >     token0 {
> >       id
> >     }
> >     token1 {
> >       id
> >     }
> >     liquidity
> >     sqrtPrice
> >     token0Price
> >     token1Price
> >     totalValueLockedUSD
> >     totalValueLockedETH
> >     totalValueLockedToken0
> >     totalValueLockedToken1
> >   }
> > }
> > ```
> 
> image
> 122 WCNDL to one USDC

ah, it is price of one token respective to other not absolute: https://thegraph.cndlchain.com/subgraphs/name/ianlapham/uniswap-v3/graphql?query=%0Aquery+get_tvl+%7B%0A++pools+%7B%0A++++id%0A++++token0+%7B%0A++++++id%0A++++%7D%0A++++token1+%7B%0A++++++id%0A++++%7D%0A++++token0Price%0A++++token1Price%0A++++totalValueLockedToken0%0A++++totalValueLockedToken1%0A++%7D%0A%7D

**Comment by samisbakedham** [comment]
> > > @samisbakedham I was trying your graph endpoint, the token count in pools are correct, but prices are like really off, like $122 for cndl token, are you looking into it?
> 
> > > ```
> 
> > > query get_tvl {
> 
> > >   
> 
> > >   pools {
> 
> > >     id
> 
> > >     token0 {
> 
> > >       id
> 
> > >     }
> 
> > >     token1 {
> 
> > >       id
> 
> > >     }
> 
> > >     liquidity
> 
> > >     sqrtPrice
> 
> > >     token0Price
> 
> > >     token1Price
> 
> > >     totalValueLockedUSD
> 
> > >     totalValueLockedETH
> 
> > >     totalValueLockedToken0
> 
> > >     totalValueLockedToken1
> 
> > >   }
> 
> > > }
> 
> > > ```
> 
> > 
> 
> > image
> 
> > 122 WCNDL to one USDC
> 
> 
> 
> ah, it is price of one token respective to other not absolute: https://thegraph.cndlchain.com/subgraphs/name/ianlapham/uniswap-v3/graphql?query=%0Aquery+get_tvl+%7B%0A++pools+%7B%0A++++id%0A++++token0+%7B%0A++++++id%0A++++%7D%0A++++token1+%7B%0A++++++id%0A++++%7D%0A++++token0Price%0A++++token1Price%0A++++totalValueLockedToken0%0A++++totalValueLockedToken1%0A++%7D%0A%7D

I think

**Comment by samisbakedham** [comment]
Merge?

**Comment by llamatester** [comment]
The adapter at projects/carthage exports TVL: 
        
 
 
    candle                    0
    
    total                    0

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    TypeError: Cannot read properties of undefined (reading 'totalValueLockedUSD')
    Truncated error stack:
    
        at tvl (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/projects/carthage/index.js:23:49)
        at async getTvl (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/test.js:59:25)
        at async /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/test.js:162:9
        at async Promise.all (index 0)
        at async Promise.all (index 2)
        at async /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/test.js:207:3

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    TypeError: Cannot read properties of undefined (reading 'totalValueLockedUSD')
    Truncated error stack:
    
        at tvl (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/projects/carthage/index.js:23:49)
        at async getTvl (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/test.js:59:25)
        at async /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/test.js:162:9
        at async Promise.all (index 0)
        at async Promise.all (index 2)
        at async /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/test.js:207:3

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    Error: invalid BigNumber string (argument="value", value="NaN", code=INVALID_ARGUMENT, version=bignumber/5.6.2)
    Truncated error stack:
        at Function.BigNumber.from (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/bignumber/lib/bignumber.js:199:27)
        at new Promise (<anonymous>)

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    Error: missing revert data in call exception; Transaction reverted without a reason string [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (transaction={"to":"0x5Bb7BAE25728e9e51c25466D2A15FaE97834FD95","data":"0x574f2ba3","accessList":null}, code=CALL_EXCEPTION, version=providers/5.6.8)
    Truncated error stack:
        at /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/providers/lib/fallback-provider.js:650:52
        at Array.forEach (<anonymous>)

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    Error: missing revert data in call exception; Transaction reverted without a reason string [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (transaction={"to":"0x5Bb7BAE25728e9e51c25466D2A15FaE97834FD95","data":"0x574f2ba3","accessList":null}, code=CALL_EXCEPTION, version=providers/5.6.8)
    Truncated error stack:
        at /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/providers/lib/fallback-provider.js:650:52
        at Array.forEach (<anonymous>)

**Comment by llamatester** [comment]
Error while running adapter at projects/carthage: 
        
 
  
    
    
    Error: missing revert data in call exception; Transaction reverted without a reason string [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (transaction={"to":"0x5Bb7BAE25728e9e51c25466D2A15FaE97834FD95","data":"0x574f2ba3","accessList":null}, code=CALL_EXCEPTION, version=providers/5.6.8)
    Truncated error stack:
        at /home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/providers/lib/fallback-provider.js:650:52
        at Array.forEach (<anonymous>)

**Comment by llamatester** [comment]
Error while running adapter at projects/candle: 
        
 
  
    
    
    Error: invalid BigNumber string (argument="value", value="NaN", code=INVALID_ARGUMENT, version=bignumber/5.6.2)
    Truncated error stack:
        at Function.BigNumber.from (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/bignumber/lib/bignumber.js:199:27)
        at new Promise (<anonymous>)

**Comment by llamatester** [comment]
Error while running adapter at projects/candle: 
        
 
  
    
    
    Error: invalid BigNumber string (argument="value", value="NaN", code=INVALID_ARGUMENT, version=bignumber/5.6.2)
    Truncated error stack:
        at Function.BigNumber.from (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/bignumber/lib/bignumber.js:199:27)
        at new Promise (<anonymous>)

**Comment by llamatester** [comment]
Error while running adapter at projects/candle: 
        
 
  
    
    
    Error: invalid BigNumber string (argument="value", value="NaN", code=INVALID_ARGUMENT, version=bignumber/5.6.2)
    Truncated error stack:
        at Function.BigNumber.from (/home/runner/work/DefiLlama-Adapters/DefiLlama-Adapters/node_modules/@ethersproject/bignumber/lib/bignumber.js:199:27)
        at new Promise (<anonymous>)

**Comment by llamatester** [comment]
The adapter at projects/carthage exports TVL: 
        
 
 
    candle                    8.80 B
    
    total                    8.80 B

File: speedTest.js (hunk 1 above)
Comment (g1nt0ki) [reply speedTest.js:24]: @samisbakedham still need to revert this change

**Comment by llamatester** [comment]
The adapter at projects/carthage exports TVL: 
        
 
 
    candle                    3.47 B
    
    total                    3.47 B

File: speedTest.js (hunk 1 above)
Comment (samisbakedham) [reply speedTest.js]: Fixed it

**Comment by g1nt0ki** [comment]
> The adapter at projects/carthage exports TVL:
> 
> ```
> candle                    3.47 B
> 
> total                    3.47 B 
> ```

this TVL looks too high

**Comment by samisbakedham** [comment]
> > The adapter at projects/carthage exports TVL:
> 
> > 
> 
> > ```
> 
> > candle                    3.47 B
> 
> > 
> 
> > total                    3.47 B 
> 
> > ```
> 
> 
> 
> this TVL looks too high

Yeah it's a glitch, shows eth price at $1 million. Fixed now

**Comment by llamatester** [comment]
The adapter at projects/carthage exports TVL: 
        
 
 
    candle                    419.33 M
    
    total                    419.33 M

**Comment by samisbakedham** [comment]
@g1nt0ki

**Comment by g1nt0ki** [comment]
> 

I am sorry but this number is also bullshit, I looked at the pools, according to the graph api, this pool is supposed to hold 100M worth of tokens: https://candleexplorer.com/address/0x6d067063e40963b9cb8F6C8576ee0c256bCf33cA

but it has 97 ETH and 18k USDC, I am not sure if I can trust those are actual eth and USDC that is bridged, what is the bridge protocol used here, so I can verify that these tokens are backed.

Literally all these pools are seeded by this wallet https://candleexplorer.com/address/0x546D090bbcEC3d96903d41e38C3436c1C601AF9c? is it candle team?

**Comment by samisbakedham** [comment]
> > 
> 
> 
> 
> I am sorry but this number is also bullshit, I looked at the pools, according to the graph api, this pool is supposed to hold 100M worth of tokens: https://candleexplorer.com/address/0x6d067063e40963b9cb8F6C8576ee0c256bCf33cA
> 
> 
> 
> but it has 97 ETH and 18k USDC, I am not sure if I can trust those are actual eth and USDC that is bridged, what is the bridge protocol used here, so I can verify that these tokens are backed.
> 
> 
> 
> Literally all these pools are seeded by this wallet https://candleexplorer.com/address/0x546D090bbcEC3d96903d41e38C3436c1C601AF9c? is it candle team? 

Yes, it's being fixed

**Comment by samisbakedham** [comment]
Fixed i think

**Comment by llamatester** [comment]
The adapter at projects/carthage exports TVL: 
        
 
 
    candle                    2.07 M
    
    total                    2.07 M

**Comment by llamatester** [comment]
The adapter at projects/carthage exports TVL: 
        
 
 
    candle                    123.36 k
    
    total                    123.36 k

**Comment by llamatester** [comment]
The adapter at projects/carthage exports TVL: 
        
 
 
    candle                    396.41 k
    
    total                    396.41 k

**Comment by g1nt0ki** [comment]
@samisbakedham how is ETH/USDC/WBTC and other other tokens bridged to candle network?

**Comment by llamatester** [comment]
The adapter at projects/carthage exports TVL: 
        
 
 
    candle                    399.59 k
    
    total                    399.59 k

**Comment by g1nt0ki** [comment]
@samisbakedham https://bridge.candlelabs.org/ is this the only bridge? It seems to be down, I want to send over some funds to test

**Comment by g1nt0ki** [comment]
sorry the more I look at the bridged tokens, the less I trust it, 600M worth of DAI and 20K ETH. All activity is from one address, can you please explain these

**Comment by samisbakedham** [comment]
> sorry the more I look at the bridged tokens, the less I trust it, 600M worth of DAI and 20K ETH. All activity is from one address, can you please explain these

Testing

**Pull request merged**